examples/riot/coap_channel/coap_channel.c

```diff
@@ -0,0 +1,284 @@
+  unsigned msg_type = COAP_TYPE_NON;
+
+  gcoap_req_init(&pdu, &buf[0], CONFIG_GCOAP_PDU_BUF_SIZE, msg_type, path);
+  coap_hdr_set_type(pdu.hdr, msg_type);
+  ssize_t len = coap_opt_finish(&pdu, COAP_OPT_FINISH_NONE);
+  ssize_t bytes_sent = gcoap_req_send(buf, len, remote, NULL, resp_handler, NULL, GCOAP_SOCKET_TYPE_UDP);
+  LF_DEBUG(NET, "CoapChannel: %d", bytes_sent);
+  if (bytes_sent > 0) {
+    LF_DEBUG(NET, "CoapChannel: Successfully sent");
+    return true;
+  }
+
+  return false;
+}
+
+static ssize_t _server_connect_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapChannel: Server connect handler");
+  CoapChannel *self = _get_coap_channel_by_remote(ctx->remote);
+  (void)self;
+
+  // TODO: If state not open or already connected, then: return bool with FALSE and also handle this bool in client
+  // TODO: Also: if self is NULL, then there is no channel that this federate supports for the other federate => return
+  // false as well.
+
+  // gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  // coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  // size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // if (self->connected_to_client) {
+  //   // This server is already connected to a client
+  //   return false;
+  // }
+
+  // // Set as connected
+  // self->connected_to_client = true;
+
+  // // Connect successful
+  // return true;
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  /* write the RIOT board name in the response buffer */
+  if (pdu->payload_len >= strlen(RIOT_BOARD)) {
+    memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
+    return resp_len + strlen(RIOT_BOARD);
+  } else {
+    puts("gcoap_cli: msg buffer too small");
+    return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+  }
+}
+
+static ssize_t _server_disconnect_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapChannel: Server disconnect handler");
+  CoapChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // Set as disconnect
+  self->state = NETWORK_CHANNEL_STATE_DISCONNECTED;
+
+  // Disconnect successful
+  return true;
+}
+
+static ssize_t _server_message_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapChannel: Server message handler");
+  CoapChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // TODO
+
+  return true;
+}
+
+static const coap_resource_t _resources[] = {
```

Comment:
This is how we register callbacks for certain incoming messages?

Reply:
This is how we register callbacks for CoAP endpoints just like HTTP endpoints.
If we want to have endpoints per message type is up for debate if it makes sense, but it is definitely doable.